### PR TITLE
Add design proposal for reliable vote transmission

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Vote transmission is critical to the function of the network.  It needs a design.

#### Summary of Changes

Proposal for reliable vote transmission.

Fixes #

### CODE REVIEW

File: book/src/reliable-vote-transmission.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+# Reliable Vote Transmission
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: Need to modify SUMMARY.md to include this

File: book/src/reliable-vote-transmission.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+drift.  So many nodes are likely to have an incorrect view if the
+next leader is active in realtime or not.
+
+2. The next leader maybe easily be flooded.  Thus a DDOS would not
+only prevent delivery of regular transactions, but also consensus
+messages.
+
+3. Udp is unreliable, and our asynchronous protocol requires any
+message that is transmitted to be retransmitted until it is observed
+in the ledger.  Potentially creating a DDOS against the leader with
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: The sentence starting with "Potentially" is not a complete sentence.

File: book/src/reliable-vote-transmission.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+a large number of validators.  Worst case flood would be Number of
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: Can you use either a notation that uses Rust conventions (`num_nodes * num_retransmits`) or English ("product of the number of nodes and number of retransmits")?

File: book/src/reliable-vote-transmission.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+critical that they are reliably delivered and encoded into the
+ledger.
+
+## Challenges
+
+1. Leader rotation is triggered by PoH, which is clock with high
+drift.  So many nodes are likely to have an incorrect view if the
+next leader is active in realtime or not.
+
+2. The next leader maybe easily be flooded.  Thus a DDOS would not
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: may be

File: book/src/reliable-vote-transmission.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+only prevent delivery of regular transactions, but also consensus
+messages.
+
+3. Udp is unreliable, and our asynchronous protocol requires any
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: UDP

File: book/src/reliable-vote-transmission.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+message that is transmitted to be retransmitted until it is observed
+in the ledger.  Potentially creating a DDOS against the leader with
+a large number of validators.  Worst case flood would be Number of
+nodes * Number of retransmits.
+
+4. Tracking if the vote has been transmitted or not via the ledger
+does not guarantee it will appear in a confirmed block.  The current
+observed block may be unrolled, validators would need to maintain
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: That comma should be a semicolon.

File: book/src/reliable-vote-transmission.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+3. Validators that receive votes in the ledger will add them to
+their local crds table, not as a push request, but simply add them
+to the table.  This shortcuts the push message protocol, so the
+validation messages do not need to be retransmitted twice around
+the network.
+
+4. CrdsValue for vote should look like this ``` Votes(Vec<Transaction>)
+```
+
+Each vote Transaction should maintain a `wallclock` in its userdata.
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: No need to capitalize transaction

File: book/src/reliable-vote-transmission.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+The merge strategy for Votes will keep the last N set of votes as
+configured by the local client.  For push/pull the vector is traversed
+recursively and each Transaction is treated as an individual CrdsValue
+with it's own local wallclock and signature.
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: its

File: book/src/reliable-vote-transmission.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+Gossip is designed for efficient propagation of state.  Messages
+that are sent through gossip push are propagated with a minimum
+spanning tree to the rest of the network, and any partial failures
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: Run-on. Split into two sentences.

File: book/src/reliable-vote-transmission.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+in the tree are actively repaired with the pull protocol while
+minimizing the amount of data transfered between any nodes.
+
+
+## How this design solves the Challenges
+
+1. Because there is no easy way for validators to be in sync with
+leaders on the leaders "active" state, gossip allows for eventual
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: the second "leaders" should be "leader's"

File: book/src/reliable-vote-transmission.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+so if the current leader is flooded the next leader would have
+already received these votes and is able to encode them.
+
+3. Gossip minimizes the number of requests through the network by
+maintaining an efficient spanning tree, and using bloom filters to
+repair state.  So retransmit back-off is not necessary and messages
+are batched.
+
+4. Leaders that read the crds table for votes will encode all the
+new valid votes that appear in the table.  Even this this leaders
```
Comment (garious) [inline book/src/reliable-vote-transmission.md]: this this -> if this?

Comment (garious) [reply book/src/reliable-vote-transmission.md]: leaders -> leader's

**Comment by aeyakovenko** [comment]
Thanks!  I’ll take a look later today.

**Pull request merged**